File: BehaviorTree/Branch/Composite/Selector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FluentBehaviorTree
{
    public class Selector : Composite
    {
        public Selector()
        {
            children = new List<Node>();
        }

        /// <summary>
        /// Propagate tick to children. Return FAILURE if no child succeeds
        /// </summary>
        /// <returns></returns>
        protected override Status tick()
        {
            if (this.Result == Status.RUNNING)
            {
                foreach (var n in children)
                {
                    if (!n.IsOpen && n.Result == Status.SUCCESS)
                    {
                        this.Result = Status.SUCCESS;
                        return this.Result;
                    }
                    else if (n.Result == Status.RUNNING)
                    {
                        this.Result = n.Tick();
                        if (this.Result != Status.FAILURE) return this.Result;
                    }
                }
                this.Result = Status.FAILURE;
            }
            return this.Result;
        }
    }
}

File: BehaviorTreesLib/Selector.cpp
#include "Selector.h"

namespace fluentBehaviorTree
{
	Selector::Selector(std::string name)
	{
		setName(name);
	}

	// Propagate tick to children.Return FAILURE if no child succeeds
	EStatus Selector::tickNode()
	{
		if (this->getResult() == EStatus::RUNNING)
		{
			for each (auto n in mChildren)
			{
				if (!n->isOpen() && n->getResult() == EStatus::SUCCESS)
				{
					this->setResult(EStatus::SUCCESS);
					return this->getResult();
				}
				else if (n->getResult() == EStatus::RUNNING)
				{
					this->setResult(n->tick());
					if (this->getResult() != EStatus::FAILURE) return this->getResult();
				}
			}
			this->setResult(EStatus::FAILURE);
		}
		return this->getResult();
	}
}